src/main/java/org/apache/sling/api/resource/path/PathSet.java

```diff
@@ -222,4 +243,223 @@ public boolean equals(final Object obj) {
+         *              contained {@link Path}s is null.
+         */
+        private void addAll(@NotNull Iterable<Path> paths) {
+            paths.forEach(this::add);
+        }
+
+        /**
+         * Adds a single {@link Path} to the {@link Builder}.
+         *
+         * @param path the <code>path</code> to be added
+         * @throws NullPointerException when path is <code>null</code>
+         */
+        private void add(Path path) {
+            if (path.isPattern()) {
+                if (patterns == null) {
+                    patterns = new HashSet<>();
+                }
+                if (patterns.add(path) && root != null) {
+                    // remove all that match the added pattern from the current tree
+                    // create a temporary copy to prevent concurrent modification of the base collection(s)
+                    Predicate<Node> condition = node -> path.matches(node.path);
+                    traverse(root, condition).filter(condition).collect(Collectors.toList()).forEach(Node::remove);
+                }
+            } else {
+                if (patterns != null && patterns.stream().anyMatch(pattern -> pattern.matches(path.getPath()))) {
+                    return;
+                }
+                if (root == null) {
+                    root = new Node("/");
+                }
+
+                Node current = root;
+                for (String part : getAncestors(path.getPath())) {
+                    // an ancestor of currently added Path is already included, skipping
+                    if (current.isIncluded()) {
+                        return;
+                    }
+                    current = current.addChild(part);
+                }
+                current.setPayload(path);
+            }
+        }
+
+        /**
+         * Returns a {@link PathSet} from the {@link Builder}'s current state,
+         * resetting the Builder afterwards.
+         *
+         * @return an optimized {@link PathSet}
+         */
+        private PathSet build() {
+            try {
+                if (patterns == null && root == null) {
+                    return PathSet.EMPTY_SET;
+                } else if (patterns != null && root == null) {
+                    // if we only have patterns the set is already optimized.
+                    // According to Path#matches(String) two patterns only
+                    // match when they are equal. For at set of patterns this
+                    // condition is always met for all elements in the Set.
+                    return new PathSet(patterns);
+                } else {
+                    Stream<Path> paths = traverse(root).filter(Node::isIncluded).map(Node::getPayload);
+                    if (patterns != null) {
+                        paths = Stream.concat(patterns.stream(), paths);
+                    }
+                    return new PathSet(paths::iterator);
```

Comment:
The Iterable's Iterator can only be consumed once. The elements need to be either collected or the stream recreated for each iterator.